src/Private/ContentHash.cs

```diff
@@ -25,8 +24,8 @@ public static async Task<string> CalculateAsync(HttpContent? content)
                 return AWS4Signer.EmptyBodySha256;
             }
 
-            var data = await content.ReadAsByteArrayAsync();
-            var hash = AWS4Signer.ComputeHash(data);
+            var contentStream = await content.ReadAsStreamAsync();
```

Comment:
@FantasticFiasco I thought of doing this, but decided it might be a downgrade from the existing code. `await ... ReadAsStreamAsync()` only waits until it can start streaming, I believe. When we pass the stream to be read by a sychronous method, it will be streamed synchronously. As such, we introduce the risk of doing part of the IO synchronously, which made the original async implementation seem superior to me.

Or do we know the response body to be always miniscule?

Reply:
It's possible this is mitigated by the fact that when the stream becomes available, some number of bytes is generally available. Might be sufficient if the body is known to be tiny.

Reply:
> await ... ReadAsStreamAsync() only waits until it can start streaming, I believe

Didn't know that, will have to read up on it. Do you have any links regarding it?

> Or do we know the response body to be always miniscule?

No we don't. The body can be of any size, and we should support them all.

Reply:
> Do you have any links regarding it?

Actually, it's good that you asked. It turns out that there is some [confusion](https://github.com/dotnet/runtime/issues/31316) about it:

- `HttpClient.GetStreamAsync()` definitely does _not_ buffer. (Not what we are interested in here, but definitely a cause of confusion if it turns out to behave differently from `HttpContent`.)
- `HttpContent.ReadAsStreamAsync()` _does_ buffer.
- However, "the CreateContentReadStreamAsync method it uses is virtual, with the intent that any self-respecting implementation overrides it to do something better" (Stephen Toub).

By the way, it took me a while to realize that we are merely reading our own output stream. That should get either buffered (usually) or re-streamed for us (such as with streaming file uploads).

Whenever we do have a buffered request body, as in the usual case, I believe I read somewhere that `ReadAsStreamAsync()` should actually reset its position and return that stream itself. That's good. That reduces copying.

In conclusion: I believe the current approach is ideal after all. The PR looks good to me.